mmeds/util.py

```diff
@@ -1654,32 +1706,74 @@ def format_table_to_lefse(i_table, metadata_file, metadata_column_class, metadat
         if metadata_column_subject:
             categories[cell][metadata_column_subject] = mdf[metadata_column_subject]['categorical'][i]
 
+    # Replace occurrences of ';' delimiter with '|'
+    if swap_delim:
+        for i, cell in enumerate(path_df[0]):
+            path_df.at[i, 0] = cell.replace(';', '|')
+
     # Insert new metadata rows into feature table
     t = [metadata_column_class]
+    to_drop = []
     for i, cell in enumerate(path_df.loc[0]):
         if i == 0:
             continue
+        if pd.isna(categories[cell][metadata_column_class]):
+            to_drop.append(i)
         t.append(categories[cell][metadata_column_class])
     # Note: using the loc[#.#] format is a bit crude, but the best way I could
     #   come up with for inserting rows without deleting already existing rows
     #   or copying each line into a new df one by one
-    path_df.loc[1.5] = t
+    path_df.loc[0.5] = t
 
     if metadata_column_subclass:
         t = [metadata_column_subclass]
         for i, cell in enumerate(path_df.loc[0]):
             if i == 0:
                 continue
+            if i not in to_drop and pd.isna(categories[cell][metadata_column_subclass]):
+                to_drop.append(i)
             t.append(categories[cell][metadata_column_subclass])
-        path_df.loc[1.6] = t
+        path_df.loc[0.6] = t
 
     if metadata_column_subject:
         t = [metadata_column_subject]
         for i, cell in enumerate(path_df.loc[0]):
             if i == 0:
                 continue
             t.append(categories[cell][metadata_column_subject])
-        path_df.loc[1.7] = t
+        path_df.loc[0.7] = t
 
     path_df = path_df.sort_index().reset_index(drop=True)
-    path_df.to_csv(o_table, sep='\t', index=False, na_rep='nan')
+    path_df = path_df.drop([0])
+    # Remove samples with a nan in class or subclass
+    if remove_nans:
+        path_df = path_df.drop(columns=to_drop)
+        Logger.debug("dropped nans")
+    path_df.to_csv(o_table, sep='\t', index=False, header=False, na_rep='nan')
+
+
+def concatenate_metadata_subsets(samples, paths):
+    """ Create a full dataframe of metadata based on the subsets of several """
+    # Create empty for concating
+    df = pd.DataFrame()
+    for p in paths:
+        # Get individual subsets
+        add_df = get_sample_subset_from_metadata(paths[p], samples[p])
+        Logger.debug(add_df)
+        df = pd.concat([df, add_df], ignore_index=True)
+    return df
+
+
+def get_sample_subset_from_metadata(metadata_file, samples, id_col=('RawData', 'RawDataID')):
+    """ Returns a dataframe of only the specified samples from a given metadata file """
+    df = pd.read_csv(metadata_file, sep='\t', header=[0, 1])
+    # Always include the additional data in the extra header rows
+    subset = [0, 1, 2]
+
+    # Append necessary rows to list
+    for i, cell in enumerate(df[id_col]):
```

Comment:
should be better way to do this with something like 

df = df[samples]
and df = df.filter(items=subset)